File: LeetCode problem set/69_Sqrt_x.js
// https://leetcode.com/problems/sqrtx/

const mySqrt = (x) => {
  if (x === 0) return x;
  let first = 1,
    last = x;
  while (first <= last) {
    let mid = Math.floor(first + (last - first) / 2);
    if (mid === x / mid) return mid;
    else if (mid > x / mid) last = mid - 1;
    else first = mid + 1;
  }
  return last;
};

console.log(mySqrt(8));

File: LeetCode problem set/69_Sqrt_x.cpp
// https://leetcode.com/problems/sqrtx/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  int mySqrt(int x) {
    if(x==0)
      return x;
    int first = 1, last = x;
    while(first <= last) {
      int mid = first + (last - first) / 2;
      if(mid == x / mid)
        return mid;
      else if(mid > x / mid)
        last = mid - 1;
      else
        first = mid + 1;
    }
    return last;
  }
};

int main() {
  Solution s;
  int x = 8;
  cout << s.mySqrt(x) << "\n";
  return 0;
}